client-mr/src/main/java/org/apache/hadoop/mapreduce/v2/app/RssMRAppMaster.java

```diff
@@ -0,0 +1,172 @@
+    int heartBeatThreadNum = conf.getInt(RSS_CLIENT_HEARTBEAT_THREAD_NUM,
+        RSS_CLIENT_HEARTBEAT_THREAD_NUM_DEFAULT_VALUE);
+    int retryMax = conf.getInt(RSS_CLIENT_RETRY_MAX, RSS_CLIENT_RETRY_MAX_DEFAULT_VALUE);
+    long retryIntervalMax = conf.getLong(RSS_CLIENT_RETRY_INTERVAL_MAX, RSS_CLIENT_RETRY_INTERVAL_MAX_DEFAULT_VALUE);
+    String coordinators = conf.get(RSS_COORDINATOR_QUORUM);
+
+    int replica = conf.getInt(RSS_DATA_REPLICA, RSS_DATA_REPLICA_DEFAULT_VALUE);
+    int replicaWrite = conf.getInt(RSS_DATA_REPLICA_WRITE, RSS_DATA_REPLICA_WRITE_DEFAULT_VALUE);
+    int replicaRead = conf.getInt(RSS_DATA_REPLICA_READ, RSS_DATA_REPLICA_READ_DEFAULT_VALUE);
+    ShuffleWriteClient client = ShuffleClientFactory
+        .getInstance()
+        .createShuffleWriteClient(clientType, retryMax, retryIntervalMax,
+            heartBeatThreadNum, replica, replicaWrite, replicaRead);
+
+    LOG.info("Registering coordinators {}", coordinators);
+    client.registerCoordinators(coordinators);
+
+    String containerIdStr =
+        System.getenv(ApplicationConstants.Environment.CONTAINER_ID.name());
+    ContainerId containerId = ContainerId.fromString(containerIdStr);
+    ApplicationAttemptId applicationAttemptId =
+        containerId.getApplicationAttemptId();
+    ShuffleAssignmentsInfo response = client.getShuffleAssignments(
+        applicationAttemptId.toString(), 0, numReduceTasks,
+        1, Sets.newHashSet(Constants.SHUFFLE_SERVER_VERSION));
+    Map<ShuffleServerInfo, List<PartitionRange>> serverToPartitionRanges = response.getServerToPartitionRanges();
+    final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+    if (serverToPartitionRanges == null || serverToPartitionRanges.isEmpty()) {
+      return;
+    }
+    LOG.info("Start to register shuffle");
+    long start = System.currentTimeMillis();
+    serverToPartitionRanges.entrySet().forEach(entry -> {
+      client.registerShuffle(
+          entry.getKey(), applicationAttemptId.toString(), 0, entry.getValue());
+    });
+    LOG.info("Finish register shuffle with " + (System.currentTimeMillis() - start) + " ms");
+    long heartbeatInterval = conf.getLong(RSS_HEARTBEAT_INTERVAL, RSS_HEARTBEAT_INTERVAL_DEFAULT_VALUE);
+    long heartbeatTimeout = conf.getLong(RSS_HEARTBEAT_TIMEOUT, heartbeatInterval / 2);
+    scheduledExecutorService.scheduleAtFixedRate(
+        () -> {
+          try {
+            client.sendAppHeartbeat(applicationAttemptId.toString(), heartbeatTimeout);
+            LOG.info("Finish send heartbeat to coordinator and servers");
+          } catch (Exception e) {
+            LOG.warn("Fail to send heartbeat to coordinator and servers", e);
+          }
+        },
+        heartbeatInterval / 2,
+        heartbeatInterval,
+        TimeUnit.MILLISECONDS);
+
+    // write shuffle worker assignments to submit work directory
+    // format is as below:
```

Comment:
What is the maximal number of partitions that can be set ?

Reply:
Integer.MAX_VALUE.

Reply:
Ok

Reply:
1 w partition cost 71ms, performance is ok.